File: 프로그래머스/3/49191. 순위/순위.kt
import java.util.*
class Solution {
    fun solution(n: Int, results: Array<IntArray>): Int {
        var ans = 0
        val p_graph = mutableMapOf<Int,MutableList<Int>>()
        for (i in 1..n) {
            p_graph.put(i,mutableListOf<Int>())
        }
        val c_graph = mutableMapOf<Int,MutableList<Int>>()
        for (i in 1..n) {
            c_graph.put(i,mutableListOf<Int>())
        }
        for (r in results) {
            val a = r[0]
            val b = r[1]
            p_graph[b]!!.add(a)
            c_graph[a]!!.add(b)
        }
        fun p_bfs(idx: Int): Int {
            var cnt = 0
            val q = ArrayDeque<Int>()
            var visit = BooleanArray(n+1) {false}
            visit[idx] = true
            q.offer(idx)
            while (q.isNotEmpty()) {
                val cur = q.poll()
                for (nxt in p_graph[cur]!!) {
                    if (!visit[nxt]) {
                        visit[nxt] = true
                        cnt++
                        q.offer(nxt)
                    }
                }
            }
            return cnt
        }
        fun c_bfs(idx: Int): Int {
            var cnt = 0
            val q = ArrayDeque<Int>()
            var visit = BooleanArray(n+1) {false}
            visit[idx] = true
            q.offer(idx)
            while (q.isNotEmpty()) {
                val cur = q.poll()
                for (nxt in c_graph[cur]!!) {
                    if (!visit[nxt]) {
                        visit[nxt] = true
                        cnt++
                        q.offer(nxt)
                    }
                }
            }
            return cnt
        }
        
        for (idx in 1..n) {
            var cnt = 1
            cnt += p_bfs(idx)
            cnt += c_bfs(idx)
            if (cnt == n) ans++
        }
        return ans
    }
}

File: 프로그래머스/3/49191. 순위/순위.py
from collections import defaultdict,deque
def solution(n, results):
    ans = 0
    # a,b -> a가 b를 이김
    p_graph = defaultdict(list)
    c_graph = defaultdict(list)
    for a,b in results:
        p_graph[b].append(a)
        c_graph[a].append(b)
    def p_bfs(i):
        q = deque()
        visit = [False]*(n+1)
        visit[i] = True
        q.append(i)
        cnt = 0
        while q:
            cur = q.popleft()
            for nxt in p_graph[cur]:
                if not visit[nxt]:
                    visit[nxt] = True
                    q.append(nxt)
                    cnt += 1
        return cnt
    def c_bfs(i):
        q = deque()
        visit = [False]*(n+1)
        visit[i] = True
        q.append(i)
        cnt = 0
        while q:
            cur = q.popleft()
            for nxt in c_graph[cur]:
                if not visit[nxt]:
                    visit[nxt] = True
                    q.append(nxt)
                    cnt += 1
        return cnt
    for i in range(1,n+1):
        # 부모와 자식 합
        cnt = 1
        cnt += p_bfs(i)
        cnt += c_bfs(i)
        if cnt == n: ans += 1
    return ans